File: cli/src/commands/vote.ts
import chalk from 'chalk'
import { requireApiKey } from '../config.js'
import { api } from '../api.js'
import { up, down } from '../theme.js'

export async function voteCommand(
  postId: string,
  options: { up?: boolean; down?: boolean; undo?: boolean }
): Promise<void> {
  const key = requireApiKey()

  let direction: number
  if (options.undo) {
    direction = 0
  } else if (options.up) {
    direction = 1
  } else if (options.down) {
    direction = -1
  } else {
    console.error(chalk.red('Specify --up, --down, or --undo'))
    process.exit(1)
  }

  await api('/vote', key, {
    method: 'POST',
    body: { postId, direction },
  })

  const msg =
    direction === 1 ? up('Upvoted.')
    : direction === -1 ? down('Downvoted.')
    : chalk.dim('Vote removed.')
  console.log()
  console.log(msg)
  console.log()
}

File: cli-rust/src/commands/vote.rs
use anyhow::{anyhow, Result};

use crate::api::ArkoraApi;
use crate::config;
use crate::theme;

pub async fn run(post_id: &str, up: bool, down: bool, undo: bool) -> Result<()> {
    let direction: i8 = if undo {
        0
    } else if up {
        1
    } else if down {
        -1
    } else {
        return Err(anyhow!("Specify --up, --down, or --undo"));
    };

    let cfg = config::load();
    let key = config::require_key(&cfg);
    let api = ArkoraApi::new(&config::api_url(&cfg), &key);

    let payload = serde_json::json!({ "postId": post_id, "direction": direction });
    api.post::<serde_json::Value>("/vote", &payload).await?;

    let msg = match direction {
        1 => theme::green("Upvoted."),
        -1 => theme::red("Downvoted."),
        _ => theme::dim("Vote removed."),
    };

    println!();
    println!("{msg}");
    println!();

    Ok(())
}
